Implemented torch::nn::Unflatten in libtorch

heitorschueroff opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42613 Implemented torch::nn::Unflatten in libtorch**
* #42563 Implemented non-named version of unflatten

Differential Revision: [D23030302](https://our.internmc.facebook.com/intern/diff/D23030302)

heitorschueroff committed:
Implemented torch::nn::Unflatten in libtorch

[ghstack-poisoned]

glaringlee commented on torch/csrc/api/src/nn/modules/linear.cpp:
```diff
@@ -79,6 +80,56 @@ Tensor FlattenImpl::forward(const Tensor& input) {
+}
+
+void UnflattenImpl::reset() {
+  named = c10::get_if<std::string>(&options.dim()) != nullptr;
```
In this case, we don't even need a boolean `named` right? 
We can use this check directly in  forward and pretty_print.

glaringlee reviewed (changes requested):
Commented, please fix the clang-tidy as well.

heitorschueroff committed:
Update on "Implemented torch::nn::Unflatten in libtorch"




[ghstack-poisoned]

heitorschueroff committed:
Update on "Implemented torch::nn::Unflatten in libtorch"




[ghstack-poisoned]

glaringlee commented on torch/csrc/api/src/nn/modules/linear.cpp:
```diff
@@ -80,19 +80,15 @@ Tensor FlattenImpl::forward(const Tensor& input) {
-UnflattenImpl::UnflattenImpl(const UnflattenOptions& options_) : options(options_) {
-  reset();
-}
+UnflattenImpl::UnflattenImpl(const UnflattenOptions& options_) : options(options_) {}
```
it seems clang-tidy follow morden c++ convention.
You need either add a move constructor:
UnflattenImpl::UnflattenImpl(UnflattenOption&& options_): options(std::move(options_){}

or replace current const ref constructor with the following one:
UnflattenImpl::UnflattenImpl(UnflattenOptions options_) : options(std::move(options_) {}

UnflattenImpl doesn't create a default move constructor, I doubt this is due to UnflattenOptions contains variant which compiler is not sure whether it is movable.

glaringlee reviewed (changes requested):
@heitorschueroff please fix clang-tidy lint error.

glaringlee commented on test/cpp/api/modules.cpp:
```diff
@@ -563,7 +563,7 @@ TEST_F(ModulesTest, Unflatten) {
   };
   
-  unflatten = Unflatten(UnflattenOptions("B", {{"B1", 2}, {"B2", 2}}));
+  unflatten = Unflatten(UnflattenOptions("B", UnflattenOptions::namedshape_t{{"B1", 2}, {"B2", 2}}));
```
Sry, I think I was not clear. This still won't work since you are still using initialize_list to initialize tuple (or pair). The initialize_list can not deduce the types within tuple (or pair).
Try explicit declare the type of tuple within initialize_list, something lie { tuple<X, Y>{"B1", 2}, tuple<X, Y>("B2", 2)}. I know this is not nice, but no other way

glaringlee commented on test/cpp/api/modules.cpp:
```diff
@@ -3532,7 +3532,7 @@ TEST_F(ModulesTest, PrettyPrintFlatten) {
   ASSERT_EQ(c10::str(Unflatten(UnflattenOptions(0, {2, 2}))),
     "torch::nn::Unflatten(dim=0, unflattened_size={2, 2})");
-  ASSERT_EQ(c10::str(Unflatten(UnflattenOptions("B", {{"B1", 2}, {"B2", 2}}))),
+  ASSERT_EQ(c10::str(Unflatten(UnflattenOptions("B", UnflattenOptions::namedshape_t{{"B1", 2}, {"B2", 2}}))),
```
Same here

heitorschueroff committed:
Update on "Implemented torch::nn::Unflatten in libtorch"




[ghstack-poisoned]

heitorschueroff committed:
Update on "Implemented torch::nn::Unflatten in libtorch"




[ghstack-poisoned]

glaringlee commented on torch/csrc/api/include/torch/nn/options/linear.h:
```diff
@@ -42,6 +43,33 @@ struct TORCH_API FlattenOptions {
+  typedef std::vector<std::pair<std::string, int64_t>> namedshape_t;
+
+  UnflattenOptions(int64_t dim, std::vector<int64_t> sizes);
+  UnflattenOptions(std::string dimname, namedshape_t namedshape);
```
Need a constructor of UnflattenOptinoos(const char*, namedshape_t) if you want to do something like UnflattenOptions("A", ......

glaringlee commented on torch/csrc/api/src/nn/options/linear.cpp:
```diff
@@ -9,5 +9,11 @@ LinearOptions::LinearOptions(int64_t in_features, int64_t out_features)
+UnflattenOptions::UnflattenOptions(int64_t dim, std::vector<int64_t> sizes)
+  : dim_(dim), sizes_(std::move(sizes)) {}
+
+UnflattenOptions::UnflattenOptions(std::string dimname, namedshape_t namedshape)
```
dim_ is primitive type, need a initial value, either initialize here or inside class doing int64_t dim_ = <init value, eg. 0> directly

glaringlee commented on test/cpp/api/modules.cpp:
```diff
@@ -3501,8 +3529,22 @@ TEST_F(ModulesTest, PrettyPrintIdentity) {
+  ASSERT_EQ(
+      c10::str(Unflatten(UnflattenOptions(
+          "B",
+          {std::tuple<std::string, int64_t>{"B1", 2},
```
tuple is used here. But nameshape_t now has pair, you can either use tuple or pair in nameshape_t, just need to make here consistent.

heitorschueroff replied:
I wonder how my tests built and passed with this, I'm fixing it

glaringlee commented on test/cpp/api/modules.cpp:
```diff
@@ -3501,8 +3529,22 @@ TEST_F(ModulesTest, PrettyPrintIdentity) {
+      c10::str(Unflatten(UnflattenOptions(
+          "B",
+          {std::tuple<std::string, int64_t>{"B1", 2},
+           std::tuple<std::string, int64_t>{"B2", 2}}))),
```
Same here

heitorschueroff committed:
Update on "Implemented torch::nn::Unflatten in libtorch"




[ghstack-poisoned]

glaringlee reviewed (approved):
LGTM now.

heitorschueroff committed:
Update on "Implemented torch::nn::Unflatten in libtorch"


Differential Revision: [D23030302](https://our.internmc.facebook.com/intern/diff/D23030302)

[ghstack-poisoned]

heitorschueroff committed:
Update on "Implemented torch::nn::Unflatten in libtorch"


Differential Revision: [D23030302](https://our.internmc.facebook.com/intern/diff/D23030302)

[ghstack-poisoned]

heitorschueroff committed:
Update on "Implemented torch::nn::Unflatten in libtorch"


Differential Revision: [D23030302](https://our.internmc.facebook.com/intern/diff/D23030302)

[ghstack-poisoned]

heitorschueroff committed:
Update on "Implemented torch::nn::Unflatten in libtorch"


Differential Revision: [D23030302](https://our.internmc.facebook.com/intern/diff/D23030302)

[ghstack-poisoned]

heitorschueroff committed:
Update on "Implemented torch::nn::Unflatten in libtorch"


Differential Revision: [D23030302](https://our.internmc.facebook.com/intern/diff/D23030302)

[ghstack-poisoned]
